### PR TITLE
Fix async init

### PR DESCRIPTION
**Pull request by arichiardi**
This patch improve and actually fixes problems that where kind of hidden by the fact that testing was not properly set to support the async computation of `read-eval-call`.

Summary in order of importance:
- Clearing the repl is now first class citizen with `replumb.core/reset-repl!`
- Brand new helper functions for querying and dissoc-ing symbols from the ast in `replumb.ast`
- Merge the `nodejs-options` and `browser-options` in `replumb.core/options`
- There is a new macro in `test-helpers.clj` that does all the work and allows you to specify tests as in:
  
  ```
  (h/read-eval-call-test e/*target-opts*
  ["(require 'clojure.set)"
  "(doc clojure.set)"]
  (let [docstring (unwrap-result @_res_)]
    (is (success? @_res_) (str _msg_ "should succeed."))  <- _msg_ & _res_ is an anaphora
    (is (valid-eval-result? docstring) (str _msg_ "should be a valid result"))
    (is (re-find #"Set operations such as union/intersection" docstring) (str _msg_ "should return valid docstring"))))
  ```
- The project has been set up to the new Figwheel way of connecting through `lein repl`

Promises are used for handling the async multiple execution of `read-eval-call`.

### CODE REVIEW

File: test/cljs/replumb/repl_test.cljs [hunk 1]
```diff
@@ -189,323 +177,315 @@ dissoc!
...
+;;     (is (= (repl/current-ns) 'cljs.user) "(load-file \"foo/load.clj\") should not change namespace"))
+;;   (_reset!_ '[foo.load]))
+
+;; was source-node-test/source-corner-cases
+(h/read-eval-call-test (assoc e/*target-opts* :load-fn! load/no-resource-load-fn!)
+  ["(require 'clojure.string)"
+   "(source clojure.string/trim)"
+   :after (repl/purge-cljs-user! '[clojure.string])]
+  (let [source-string (unwrap-result @_res_)]
+    (println "ARSARSARSASR")
```
Comment (tomasz-biernacki) [inline test/cljs/replumb/repl_test.cljs]: Maybe to delete

File: test/cljs/replumb/repl_test.cljs [hunk 2]
```diff
@@ -189,323 +177,315 @@ dissoc!
...
+    (is (= "nil" source-string) (str _msg_ "when *load-fn* returns nil should return nil"))))
+
+(h/read-eval-call-test (-> e/*target-opts*
+                           (dissoc :load-fn!)
+                           (assoc :read-file-fn! nil))
+  ["(require 'clojure.string)"
+   "(source clojure.string/trim)"
+   :after (repl/purge-cljs-user! '[clojure.string])]
+  (let [source-string (unwrap-result @_res_)]
+    (println "--------------------\nARSARSARSASR\n")
```
Comment (tomasz-biernacki) [inline test/cljs/replumb/repl_test.cljs]: As above.

Comment (arichiardi) [inline test/cljs/replumb/repl_test.cljs]: Intense debugging are you checking the unified already?

On March 15, 2016 4:34:41 PM PDT, Tomek <notifications@github.com> wrote:
>> +   "(source clojure.string/trim)"
>> +   :after (repl/purge-cljs-user! '[clojure.string])]
>> +  (let [source-string (unwrap-result @_res_)]
>> +    (println "ARSARSARSASR")
>> +    (is (success? @_res_) (str _msg_ "when *load-fn* returns nil
>should succeed."))
>> +    (is (valid-eval-result? source-string) (str _msg_ "when
>*load-fn* returns nil should be a valid result"))
>> +    (is (= "nil" source-string) (str _msg_ "when *load-fn* returns
>nil should return nil"))))
>> +
>> +(h/read-eval-call-test (-> e/*target-opts*
>> +                           (dissoc :load-fn!)
>> +                           (assoc :read-file-fn! nil))
>> +  ["(require 'clojure.string)"
>> +   "(source clojure.string/trim)"
>> +   :after (repl/purge-cljs-user! '[clojure.string])]
>> +  (let [source-string (unwrap-result @_res_)]
>> +    (println "--------------------\nARSARSARSASR\n")
>
>As above.
>
>---
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/Lambda-X/replumb/pull/162/files/2cff458e1a1e8fa3d0707cb60960164c3af940dc#r56259831

-ar
